docs/reference/c++_support.rst

```diff
@@ -0,0 +1,173 @@
+``__device__``, ``__shared__``, ``__managed__``, and ``__constant__`` specifiers.
+
+The ``__device__`` and ``__constant__`` specifiers define static variables, which are
+allocated within global memory on the HIP devices. The only difference is that
+``__constant__`` variables can't be changed after allocation. The ``__shared__``
+specifier allocates the variable within shared memory, which is available for all threads
+in a block.
+
+The ``__managed__`` variable specifier creates global variables that are initially
+undefined and unaddressed within the global symbol table. The HIP runtime allocates
+managed memory and defines the symbol when it loads the device binary. A managed variable
+can be accessed in both device and host code. To register managed variables, use
+``__hipRegisterManagedVariable`` in an initialization function.
+
+It's important to know where a variable is stored because it is only available from
+certain locations. Generally, variables allocated in the host memory are not accessible
+from the device code, while variables allocated in the device memory are not accessible
+from the host code. Dereferencing a pointer to device memory on the host results in a
+segmentation fault.
+
+Exception handling
+-------------------------------------------------------------------------------
+
+An important difference between the host and device code is exception handling. In device
+code, this control flow isn't available due to the hardware architecture. The device
+code must use return codes to handle errors.
+
+Kernel parameters
+-------------------------------------------------------------------------------
+
+There are some restrictions on kernel function parameters. They cannot be passed by
+reference, because these functions are called from the host but run on the device. Also,
+a variable number of arguments is not allowed.
+
+Classes
+-------------------------------------------------------------------------------
+
+Classes work on both the host and device side, but there are some constraints. The static
+data members need to be ``const`` qualified, and ``static`` member functions can't be
+``__global__``. ``Virtual`` member functions work, but a ``virtual`` function must not be
+called from the host if the parent object was created on the device, or the other way
+around, because this behavior is undefined. This also means you can't pass an object with
+``virtual`` functions as a parameter to a kernel.
+
+Polymorphic function wrappers
+-------------------------------------------------------------------------------
+
+HIP doesn't support the polymorphic function wrapper ``std::function``, which was
+introduced in C++11.
+
+Extended lambdas
+-------------------------------------------------------------------------------
+
+HIP supports Lambdas, which by default work as expected.
+
+Lambdas inherit the execution space specification from the surrounding context. For
```

Comment:
A lambda has implicit host device attributes, therefore can be executed by both device and host code. A lambda defined in a host function can be passed to a template kernel as a template argument and called in the kernel. A lambda can access host variables by copy but not by reference. Accessing host variables by reference in device code results in compilation error or undefined behavior.